File: src/tictac/TicTacToe.java
package tictac;

import java.util.Scanner;

public class TicTacToe {
	String[][] board =  {{" ", "|", " ", "|", " "},
					   {"-", "+", "-", "+", "-"},
					   {" ", "|", " ", "|", " "},
					   {"-", "+", "-", "+", "-"},
					   {" ", "|", " ", "|", " "},};
	private String p1, p2;
	
	public TicTacToe() {
		p1 = "X";
		p2 = "O";
	}
	
	public void play() {
		String player = p1;
		int moves = 0;
		
		do {
			displayBoard();
			move(player);
			if (player == p1) {
				player = p2;
			} else {
				player = p1;
			}
		} while (moves <= 9 && winner() == " ");
		displayBoard();
		System.out.println("Winner: " + winner());
		
	}
	
	private void displayBoard() {
		for(int row = 0; row < board.length; row++) {
			for(int col = 0; col < board.length; col++) {
				System.out.print(board[row][col]);
			}
			System.out.println();
		}
	}
	
	private void move(String player) {
		Scanner input = new Scanner(System.in);
		boolean valid = false;
		int row, col;
		
		do {
			System.out.println("Enter row (0-2): ");
			row = input.nextInt();
			System.out.println("Enter column (0-2): ");
			col = input.nextInt();
			if ((row >= 0 && row <= 2 && col >= 0 && col <= 2) && 
				 board[row * 2][col * 2].equals(" ")) {
				
				board[row * 2][col * 2] = player;
				valid = true;
			} else {
				System.out.println("Invalid move.");			
				
			}
		} while (!valid);
	}
	
	private String winner() {
		for(int row = 0; row <= 2; row++) {
			if (board[row * 2][0].equals(board[row * 2][2]) && 
			    board[row * 2][2].equals(board[row * 2][4]) &&
			    !(board[row * 2][0].equals(" "))) {
				
				return(board[row * 2][0]);
			}
		}
		
		for(int col = 0; col <= 2; col++) {
			if (board[0][col * 2].equals(board[2][col * 2]) && 
			    board[2][col * 2].equals(board[4][col * 2]) &&
			    !(board[0][col * 2].equals(" "))) {
				
				return(board[0][col * 2]);
			}
		}
		
		if (board[0][0].equals(board[2][2]) && 
		    board[2][2].equals(board[4][4]) && !(board[0][0].equals(" "))) {

			return(board[0][0]);
		}
		 
		if (board[0][4].equals(board[2][2]) && 
			board[2][2].equals(board[4][0]) &&
			!(board[0][4].equals(" "))) {
			
			return(board[0][4]);
		}
		return(" ");
	}
}

File: tictactoe.py
board = [[' ', '|', ' ', '|',' '],
        ['-','+','-','+','-'],[' ', '|', ' ', '|',' '],
        ['-','+','-','+','-'],[' ', '|', ' ', '|',' ']]

p1 = 'x'
p2 = 'o'

'''
Prints board
'''
def printBoard(board):
    for line in board:
        for item in line:
            print(item, end=" ")
        print()

# Checks if user entry is valid
def checkValid(row, col):
    if row < 0 or row > 2 or col < 0 or col > 2:
        print("Invalid entry")
        return False
    if board[row*2][col*2] != " ":
        print("Already taken")
        return False
    return True
        
'''
Take user input to begin game
'''
def playerMove(player):
    x = int(input("Enter row (0-2): "))
    y = int(input("Enter column (0-2): "))
    valid = checkValid(x, y)
    if not valid:
        playerMove(player)
    else:
        board[x*2][y*2] = player

def checkWin(player):
    #checks each row for win
    for row in range(3):
        if (board[row * 2][0] == board[row * 2][2] == 
            board[row * 2][4] == player):
            return True           
    
    #check columns 
    for col in range(3):
        if (board[0][col * 2] == board[2][col * 2] == 
            board[4][col * 2] == player):  
            return True

    #check diagonals
    if (board[0][0] == board[2][2] == 
        board[4][4] == player):
        return True
    return False

    
'''
Game Play
'''
def tictactoe():
    player = p1
    moves = 0
    win = False
    printBoard(board)
    for i in range(10):
        playerMove(player)
        printBoard(board)
        moves += 1
        if moves >= 5:
            win = checkWin(player)
        if win:
            print("Player "+player+" wins!")
            break
        if (player == p1):
            player = p2
        else:
            player = p1
    
    if moves == 9:
        print("Tie game")




if __name__ == "__main__":
  tictactoe()
